Add blog post on FerretDB beta release 0.8.0 (#1815)
FerretDB/FerretDB: website/blog/2023-01-04-ferretdb-beta-version-authentication.md

---
title: FerretDB v0.8.0 - The Beta Version
slug: ferretdb-beta-version-authentication
image: ../static/img/blog/FerretDB-is-now-Beta.-1-980x551.png
description: The FerretDB beta version (v.0.8.0) includes exciting new features, including authentication for PostgreSQL, `$min` operator support, and much more.
author: Alexander Fashakin
author_title: FerretDB Team
---

![FerretDB 0.8.0 release](../static/img/blog/FerretDB-is-now-Beta.-1-980x551.png)

<!--truncate-->

> FerretDB - the open source alternative to MongoDB - is thrilled to announce release of our  Beta version(0.8.0), which includes various new features, bug fixes, improved documentation, and, most importantly, the implementation of authentication for PostgreSQL.

Reviewer: ```suggestion
FerretDB - the open source alternative to MongoDB - is thrilled to announce release of our Beta version (0.8.0), which includes various new features, bug fixes, improved documentation, and, most importantly, the implementation of authentication for PostgreSQL.
```

This new and exciting release is no ordinary milestone; it's a culmination of FerretDB's journey as we work on bringing you the ultimate open-source alternative to MongoDB by converting MongoDB protocol queries to SQL, with PostgreSQL as the database engine.

While we do not aim to cover all the features of MongoDB, our goal with the Beta - and our upcoming GA release later this quarter - is to provide a solid foundation on which to build targeted features that'll enable FerretDB to support more and more real-world use cases.
Be sure to [check our roadmap for further details](https://github.com/orgs/FerretDB/projects/2).

:::caution
Please note that this particular change breaks backward compatibility.
To dump and restore the data, you’ll need connection strings to connect to FerretDB and PostgreSQL.
There are numerous ways to dump and restore your data.
For example, you can follow the following steps:

1. Backup FerretDB databases with `mongodump`.

> Set your FerretDB connection string in `—uri` and run:

Reviewer 2: ```suggestion
Set your FerretDB connection string in `--uri` and run:
```
It's rendered weirdly.

   `mongodump --uri="mongodb://localhost:27017"`
   This command will create a directory with dumps for each FerretDB database on the given server.
Later, we will use this dump to restore the database.
Instead of `mongodump`, you can also use `mongoexport` for this.

2. Backup PostgreSQL `ferretdb` database.
If the migration goes well, we will not use this backup.
But we recommend to do this step in case you need to rollback.
Specify your host and port.
`pg_dump -h localhost -p 5432 -U username ferretdb > ferretdb.sql`

3. Stop FerretDB (This depends on your operating system and the way you run FerretDB)

> 4. Connect to PostgreSQL and drop `ferredb` database as it's not needed anymore:

Reviewer 3: ```suggestion
4. Connect to PostgreSQL and drop `ferretdb` database as it's not needed anymore:
```

That was my typo 😅

   * `psql -h localhost -p 5432 -U username postgres`
   * `DROP DATABASE ferretdb`

> 5. Upgrade FerretDB and run FerretDB 0.8 (Please refer to our documentation where we describe how to update and start FerretDB)

Reviewer 3: Do we have a link or something?

6. Restore database using `mongorestore --uri="mongodb://localhost:27017"`

7. In case something doesn't work and you need to rollback to FerretDB 0.7.1:
   * Stop ferretdb
   * Delete ferretdb database (repeat step 4)
   * Restore `ferretdb` PostgreSQL db from the dump we created on the step 2:  `psql -h localhost -p 5432 -U username ferretdb -f ferretdb.sql`
   * Start ferretdb 0.7.1

> :::

Reviewer 3: In principle, in this case it should be possible (and better) to rollback with `mongorestore` too. Postgres' backup is a backup-backup, in case if something went really wrong.

Author: Is this something we want to add? "You can rollback with `mongorestore`. However, in case something doesn't work..." and then continue from there

Reviewer 3: Yes, that sounds good to me!

In this blog post, we'll be sharing detailed information about the FerretDB beta release (0.8.0).

## What's new?

Since the last release (0.7.2), we've made great strides in adding new features to FerretDB, especially the introduction of authentication for PostgreSQL.
With authentication, users can now connect securely using passwords, ensuring that only authorized connections are established.

> You can implement an authenticated connection by specifying the connection string in this format: `postgres://username:password@postgres:5432/ferretdb`.

Reviewer 3: This is how FerretDB connects to postgres, is it what we want to set in the docs?

Author: Sorry, I don't understand what you mean here.

Reviewer 3: I mean, why do we mention PostgreSQL connection string in this sentence? What do we expect people do with it?

Author: Ok. I was mostly citing the PostgreSQL connection string example for authentication in [our documentation here ](https://docs.ferretdb.io/security/#authentication). Mostly we want to have examples/use cases for each new features in the release. What do you suggest?

Reviewer 3: For example, we can add a whole sentence from the documentation (both PostgreSQL and FerretDB connection strings), as PostgreSQL one without any explanation is hard to interpret.

Author: ```suggestion
You can implement an authenticated connection to the PostgreSQL database by specifying the default username and password in the FerretDB connection string as `postgres://username:password@postgres:5432/ferretdb`,  or use a MongoDB URI `mongodb://username:password@ferretdb:27018/?tls=true&authMechanism=PLAIN`
```

Is this okay?

See more details [in our documentation](https://docs.ferretdb.io/security/#authentication).

But that's not all - in addition to the `$max` update operator, FerretDB beta now includes support for the `$min` update operator:

```js
db.collection.update( {}, { $min: { <field1>: <value1>, ... } } )
```

We've also added support for `ordered` inserts, which allows you to insert data in the exact order it comes in.

```js

> db.collection.insert( { "field1": "value1" }, { ordered: true } )

Reviewer 3: Those examples are not consistent: first one uses `<field1>: <value1>` "pseudocode", and the second one is a particular example. Should we make them written in the same way? 

Based on how they render, maybe multi-line could be more readable.

Author: Good spotting! I'll work on that now

Author: PTAL

Reviewer 3: LGTM

```

## Bug fixes and enhancements

This release includes a bug fix that addresses an issue with the `$inc` operator when used with `unset` documents.
We found that, in certain cases, an invalid value of the `$inc` operator was updating `unset` documents with non-numeric values when it should have instead returned an error.
This release should prevent unexpected updates on `unset` documents, and likewise, improve the overall stability and reliability of FerretDB.

The release includes one enhancement, which is updating our building documentation.
We have documented our build process and Go build tags for those who use embedded packages or build FerretDB themselves.
Please [check here](https://github.com/FerretDB/FerretDB/blob/main/README.md#building-and-packaging) for further details.

## Documentation

Our documentation is not left out on the list of improvements!
Specifically, we've updated it to include descriptions and examples for element query operators, array query operators, and comparison and logical query operators.

> These changes help make our documentation more comprehensive and user-friendly, which should hugely benefit our community and users alike.

Reviewer 3: Maybe in such sentences, we can ask for feedback? Like "don't hesitate to let us know what you think about it" or something :)

However, this change is not a cause for concern, merely a signal that we will be focusing on other features and improvements that are of greater priority in [our roadmap](https://github.com/orgs/FerretDB/projects/2).
This change will not affect the support for the `$elemMatch` query operator.
For more information on the supported operators and commands, please check out our [updated documentation](https://docs.ferretdb.io/reference/supported_commands/).

Please find more details about the new features and changes in the beta version [here](https://github.com/FerretDB/FerretDB/releases/tag/v0.8.0).

## And to our amazing community and users

Once again, we want to take a moment to express our gratitude to everyone who has been part of the FerretDB journey so far - either as a user, contributor, supporter, or admirer.
Not to mention the growing number of partnerships, compatible applications, and all the contributions from our incredible community.
In 2022, we had:

* 👨🏻‍💻 Over 40 code contributors with more than 130 merged pull requests from our community of contributors

> * ⭐️ 5.1k Stars on GitHub(Have you given us a Star yet? Be sure to [Star us on GitHub](https://github.com/FerretDB/FerretDB).)

Reviewer 3: ```suggestion
* ⭐️ 5.1k Stars on GitHub. Have you given us a Star yet? Be sure to [Star us on GitHub](https://github.com/FerretDB/FerretDB)
```

Or punctuation is missing around `(` and `)`

* ⏫ More than 100 Docker image downloads

We appreciate your continued and invaluable support and feedback, as we strive to make FerretDB even better.
Stay tuned for more exciting developments and updates from us!

> Please remember, if you have questions about FerretDB, feel free to [contact us](https://docs.ferretdb.io/intro/#community).

Reviewer 2: The link doesn't exist, I think this one is what we want.
```suggestion
Please remember, if you have questions about FerretDB, feel free to [contact us](https://docs.ferretdb.io/#community).
```

6. Restore database using `mongorestore --uri="mongodb://localhost:27017"`

7. In case something doesn't work and you need to rollback to FerretDB 0.7.1:
   * Stop ferretdb
   * Delete ferretdb database (repeat step 4)
   * Restore `ferretdb` PostgreSQL db from the dump we created on the step 2:  `psql -h localhost -p 5432 -U username ferretdb -f ferretdb.sql`
   * Start ferretdb 0.7.1
:::

In this blog post, we'll be sharing detailed information about the FerretDB beta release (0.8.0).

## What's new?

Since the last release (0.7.2), we've made great strides in adding new features to FerretDB, especially the introduction of authentication for PostgreSQL.
With authentication, users can now connect securely using passwords, ensuring that only authorized connections are established.
You can implement an authenticated connection by specifying the connection string in this format: `postgres://username:password@postgres:5432/ferretdb`.
See more details [in our documentation](https://docs.ferretdb.io/security/#authentication).

But that's not all - in addition to the `$max` update operator, FerretDB beta now includes support for the `$min` update operator:

```js

> db.collection.update( {}, { $min: { &lt;field1>: &lt;value1>, ... } } )

Reviewer 2: Need to update above line, formatting is odd.

## Bug fixes and enhancements

This release includes a bug fix that addresses an issue with the `$inc` operator when used with `unset` documents.
We found that, in certain cases, an invalid value of the `$inc` operator was updating `unset` documents with non-numeric values when it should have instead returned an error.
This release should prevent unexpected updates on `unset` documents, and likewise, improve the overall stability and reliability of FerretDB.

The release includes one enhancement, which is updating our building documentation.
We have documented our build process and Go build tags for those who use embedded packages or build FerretDB themselves.
Please [check here](https://github.com/FerretDB/FerretDB/blob/main/README.md#building-and-packaging) for further details.

## Documentation

Our documentation is not left out on the list of improvements!
Specifically, we've updated it to include descriptions and examples for element query operators, array query operators, and comparison and logical query operators.

These changes help make our documentation more comprehensive and user-friendly, which should hugely benefit our community and users alike.

## Other changes

> In this latest release, we've made some changes to our supported features, which include discontinuing support for $elemMatch and $slice projection operators due to existing technical issues.

Reviewer 2: ```suggestion
In this latest release, we've made some changes to our supported features, which include discontinuing support for `$elemMatch` and `$slice` projection operators due to existing technical issues.
```

This change will not affect the support for the `$elemMatch` query operator.
For more information on the supported operators and commands, please check out our [updated documentation](https://docs.ferretdb.io/reference/supported_commands/).

Please find more details about the new features and changes in the beta version [here](https://github.com/FerretDB/FerretDB/releases/tag/v0.8.0).

## And to our amazing community and users

Once again, we want to take a moment to express our gratitude to everyone who has been part of the FerretDB journey so far - either as a user, contributor, supporter, or admirer.
Not to mention the growing number of partnerships, compatible applications, and all the contributions from our incredible community.
In 2022, we had:

* 👨🏻‍💻Over 40 code contributors with more than 130 merged pull requests from our community of contributors
* ⭐️5.1k Stars on GitHub(Have you given us a Star yet? Be sure to S[tar us on GitHub](https://github.com/FerretDB/FerretDB).)
* ⏫More than 100 Docker image downloads

> * 🎤And, we actively participated in several events, including Percona Live,

Reviewer 3: Oh, I like those emojis! Should we have a whitespace between an emoji and the rest of the text?

> * 🎤 And, we actively participated in several events, including Percona Live,

Reviewer 4: Stats are great, but event participation has nothing to do with the release, I would leave it out
